pkg/hubble/metrics/dynamic_flow_processor.go

```diff
@@ -0,0 +1,134 @@
+
+	// "github.com/cilium/cilium/pkg/hubble/metrics"
+	"github.com/cilium/cilium/pkg/hubble/metrics/api"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// DynamicFlowProcessor represents instance of hubble exporter with dynamic
+// configuration reload.
+type DynamicFlowProcessor struct {
+	logger  logrus.FieldLogger
+	watcher *metricConfigWatcher
+	// mutex protects from concurrent modification of managedFlowProcessors by config
+	// reloader when hubble events are processed
+	mutex    lock.RWMutex
+	Metrics  *api.Handlers // TODO add getnames for testing encapsulation?
+	registry *prometheus.Registry
+}
+
+// OnDecodedEvent distributes events across all managed exporters.
+func (d *DynamicFlowProcessor) OnDecodedFlow(ctx context.Context, flow *flowpb.Flow) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, nil
+	default:
+	}
+
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+
+	var errs error
+	if d.Metrics != nil {
+		d.Metrics.ProcessFlow(ctx, flow)
+	}
+
+	if errs != nil {
+		d.logger.WithError(errs).Error("Failed to ProcessFlow in metrics handler")
+	}
+	return false, errs
+}
+
+// NewDynamicFlowProcessor creates instance of dynamic hubble flow exporter.
+func NewDynamicFlowProcessor(reg *prometheus.Registry, logger logrus.FieldLogger, configFilePath string) *DynamicFlowProcessor {
+	dynamicFlowProcessor := &DynamicFlowProcessor{
+		logger:   logger,
+		registry: reg,
+	}
+	watcher := NewMetricConfigWatcher(configFilePath, dynamicFlowProcessor.onConfigReload)
+	dynamicFlowProcessor.watcher = watcher
+	return dynamicFlowProcessor
+}
+
+func (d *DynamicFlowProcessor) onConfigReload(ctx context.Context, isSameHash bool, hash uint64, config api.Config) {
+	if isSameHash {
+		return
+	}
+
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	var newHandlers api.Handlers
+	metricNames := config.GetMetricNames()
+
+	curHandlerMap := make(map[string]api.NamedHandler)
+	if d.Metrics != nil {
+		for _, m := range d.Metrics.Handlers {
+			curHandlerMap[m.Name] = m
+		}
+
+		configuredMetricNames := make(map[string]*api.MetricConfig)
+		for _, cm := range config.Metrics {
+			configuredMetricNames[cm.Name] = cm
+		}
+		// Unregister handlers not present in the new config.
+		// This needs to happen first to properly check for conflicting plugins later during registration.
+		for _, m := range d.Metrics.Handlers {
+			if _, ok := configuredMetricNames[m.Name]; !ok {
+				h, _ := curHandlerMap[m.Name]
+				h.Handler.Deinit(d.registry)
+				delete(curHandlerMap, m.Name)
+			}
+		}
+	}
+
+	for _, v := range curHandlerMap {
+		newHandlers.Handlers = append(newHandlers.Handlers, v)
+	}
+
+	for _, cm := range config.Metrics {
+		// Existing handler matches new config entry:
+		//   no-op, if config unchanged;
+		//   deregister and re-register, if config changed.
+		if m, ok := curHandlerMap[cm.Name]; ok {
+			if reflect.DeepEqual(*m.MetricConfig, *cm) {
+				continue
+			} else {
+				if h, ok := curHandlerMap[cm.Name]; ok {
+					h.Handler.Deinit(d.registry)
```

Comment:
Can you elaborate this?

Reply:
What I mean is the API here might be wrong. Instead of `Deinit()`, it might make more sense to change to something like `HandleConfigurationUpdate(...)` instead.

Reply:
1. Making sure you meant just the update scenario here - I think `Deinit()` still makes sense for removing a metric.

2. Whether `metrics are stateful data we don't want to reset ideally.` could depend on which scenario will be more common: 
- user applies a filter on a metric to remove noise from uninteresting pods: here we could free up resources by removing old metrics by re-registering the handler
- user only temporarily applies a filter and removes it later on, wants to see metric state from before

3. I can add `HandleConfigurationUpdate` - since the PR doesn't support label changes, this function would basically change the filter fields on the handlers, right?

Reply:
If you don't want to handle label updates in this PR, that's fine, but I still believe updates should be treated differently from deinit + reinit. In general, I want to bias towards not re-initializing the metrics. 

Since this PR won't support label updates, it will also need to add validation to prevent attempted updates to labels.

Reply:
Validation function update, test case added